src/subsystems/nodejs/discoverers/default/default.nix

```diff
@@ -46,16 +46,27 @@
         && (packageJson.devDependencies or {} == {})
         && (packageJson.workspaces or [] == [])
       then ["package-lock"]
+      else if nodes ? "package-lock.json"
+      then let
+        # Wish there was a way to get the version without reading a 2MB file
+        lockJson = getLockJson path;
+        lockVersion = lockJson.lockfileVersion or 0;
+      in
+        if lockVersion == 1
+        then ["package-lock"]
+        else if lockVersion == 2
+        then ["package-lock-v2"]
+        else ["package-json"]
```

Comment:
The idea of the `translators` attribute is to list all possible translators.
If there is a package-lock.json and a yarn.lock, then both translators should be listed.

Also I think the new translator should probably be re-named to package-lock-v3.
The `v2` file format just contains definitions for `v1` and `v3` parsers. There isn't really a `v2` parser.
See https://docs.npmjs.com/cli/v8/configuring-npm/package-lock-json#lockfileversion

Therefore if lock file version is `2`, both `package-lock` and `package-lock-v3` translators should be listed.
dream2nix will then pick the first on of these by default, so if we want v3 to be the preferred logic, it should be the first item in the list.

Reply:
I think it would be better to make the v3 parser call the v1 parser if the version is wrong. Then the discoverer doesn't need to parse the json twice and the package-json impure translator can still use old nodejs versions. I'll give that a shot.

Reply:
I think it would be better for UX and debugging purposes if we could show the user specifically which translator implementation was used.

Actually parsing the json twice can already be prevented by making use of the `tree` object which is given to the discoverer. `tree` is created by `lib.prepareSourceTree` and presents a directory tree as an attribute set. The attribute set also contains the json parsed content of each file. Because both discoverer and translator use the same tree object, json files will only be parsed once.

Instead of `fromJSON` the discoverer should just use `(tree.getNodeFromPath "/rel/path/to/package.json").jsonContent`.

Is it really necessary to have the impure package json translator work with older nodejs versions? Couldn't we just fix the version?

Reply:
Good tips, thanks, I'll work with those!

> Is it really necessary to have the impure package json translator work with older nodejs versions? Couldn't we just fix the version?

I don't want to force a download of a specific version, but maybe I can check if the version of the nodejs pkg is >= 16 and otherwise use v18.

Reply:
@DavHau So, the pure symlink builder needs to have all peer deps and all circular deps for each dep, and neither package-lock v1 nor yarn v1 have that information.

The sure only way around that is to build the node_modules by first copying all sources into one place, and then running builds for every package. This is inefficient, with easily over 1000 dependencies sometimes. It's also tricky, clashing dependencies have to be detected and moved down into the tree.

The current builder+translator combo doesn't quite do that and fails in some cases (documented elsewhere).

For simplicity, I'd like to remove support for npm v1 and yarn v1 lockfiles. Then the pure builder has all the correct information for building a pure tree of symlinks. The disadvantage is that some projects will first need an impure npm step.

Reply:
> For simplicity
 
If you isolate your features into a new builder/translator pair, I don't see how removing existing builders/translators simplifies things for you. We only need an interface to allow the user to chose between builders translators (which already exists).

> I'd like to remove support for npm v1 and yarn v1 lockfiles

Users are currently depending on these modules. Removing that causes a lot of friction as it is unlikely that the new builders/translators will be bug free from the beginning. It will take some time to stabilize the new approach, and for that period, the old modules will still be useful.

I'd suggest the following upgrade path:
- add new logic as new translator/builder including an example under `/examples` on how to use it. (we can merge this before being fully stable, as it is an optional feature)
- generate an index of the most popular 5000 nodejs packages (similar to dream2nix-npm). I'm happy to do take that task once we merged your new transaltor/builder.
- build the 5000 packages and fix bugs.
- once stable, make the new translator/builder the default